syntax = "proto3";

package proto;

message TweetStatusP {

    string createdAt = 1;
    int64 id = 2;
    string idStr=3;
    string text = 4;
    string source = 5;
    bool isTruncated = 6;
    int64 inReplyToStatusId = 7;
    int64 inReplyToUserId = 8;
    string inReplyToScreenName = 9;
    UserP userP = 10;
    CoordinatesP coordinates = 11;
    PlaceP place = 12;
    int64 quotedStatusId = 13;
    bool isQuoteStatus = 14;
    TweetStatusP quotedStatus = 15;
    TweetStatusP retweetedStatus = 16;
    int32 quoteCount=17;
    int32 replyCount=18;
    int32 retweetCount=19;
    int32 favoriteCount=20;
    EntitiesP entitiesP=21;
    ExtendedTweetEntityP extendedTweet=22;
    bool isFavorited=23;
    bool isRetweeted=24;
    bool isPossiblySensitive=25;
    string filterLevel=26;
    string lang=27;
    repeated MatchingRulesEntityP matchingRules=28;
    int64 currentUserRetweetedId=29;
    repeated MapFieldEntry scopes=30;
    bool withheldCopyright=31;
    repeated string withheldInCountries=32;
    string  withheldScope=33;
    repeated int32 displayTextRange=34;
}

//User:
message UserP {
    int64 id = 1;
    string idStr=2;
    string name = 3;
    string screenName = 4;
    string location = 5;
    string url = 6;
    string description = 7;
    bool isProtected = 8;
    bool isVerified = 9;
    int32 followersCount = 10;
    int32 friendsCount = 11;
    int32 listedCount = 12;
    int32 favouritesCount = 13;
    int32 statusesCount = 14;
    string createdAt = 15;
    string profileBannerImageUrl = 16;
    string profileImageUrlHttps = 17;
    bool isDefaultProfile = 18;
    repeated string withheldInCountries = 19;
    string withheldScope = 20;
    URLEntityP descriptionURLEntities = 21;
    bool isGeoEnabled = 22;
    string lang = 23;
    bool isContributorsEnabled = 24;
    string profileBackgroundColor = 25;
    string profileBackgroundImageUrl = 26;
    string profileBackgroundImageUrlHttps = 27;
    bool profileBackgroundTiled = 28;
    string profileImageUrl = 29;
    string profileLinkColor = 30;
    string profileSidebarBorderColor = 31;
    string profileSidebarFillColor = 32;
    string profileTextColor = 33;
    bool profileUseBackgroundImage = 34;
    bool isDefaultProfileImage = 35;
    int32 utcOffset = 36;
    string timeZone = 37;
    bool translator = 38;
    bool isFollowRequestSent = 39;
    bool showAllInlineMedia = 40;
}

//Coordinates
message CoordinatesP {
    string type = 1;
    repeated double coordinates = 2;
}

//Place
message PlaceP {
    string name = 1;
    string countryCode = 2;
    string id = 3;
    string country = 4;
    string placeType = 5;
    string url = 6;
    string fullName = 7;

    //BoundingBoxCoordinate
    message BoundingBoxCoordinateP {
        string type = 1;

        // Representation of a three dimensional array for coordinates
        message RowsOneP {
            message RowsTwoP {
                repeated double coordinate = 1;
            }
            repeated RowsTwoP rowsTwoP = 1;
        }
        repeated RowsOneP threeDimArray = 2;

    }

    BoundingBoxCoordinateP boundingBoxCoordinateP = 8;
}

//Entities
message EntitiesP{
    repeated HashtagEntityP hashtagEntityP=1;
    repeated MediaEntityP mediaEntityP=2;
    repeated URLEntityP urlEntityP = 3;
    repeated UserMentionEntityP userMentionEntityP = 4;
    repeated SymbolEntityP symbolEntityP = 5;
    repeated PollEntityP pollEntityP = 6;

}

//ExtendedEntities
message ExtendedEntitiesP{
    repeated MediaEntityP media=1;
}

//URLEntityP
message URLEntityP {
    repeated int32 indices = 1;
    string displayURL = 2;
    string expandedURL = 3;
    string url = 4;
}

//HashtagEntity
message HashtagEntityP{
    repeated int32 indices=1;
    string text=2;
}
//MediaEntity
message MediaEntityP{
    string displayURL=1;
    string expandedURL=2;
    int64 id=3;
    string idStr=4;
    repeated int32 indices=5;
    string mediaURL=6;
    string mediaURLHttps=7;
    MediaSizesEntityP sizes=8;
    string type=9;
    int64 sourceStatusId=10;
    string sourceStatusIdStr=11;
    string url=12;
    VideoEntityP videoInfo=13;
    AdditionalMediaInfoEntityP additionalMediaInfo=14;
}

//UserMentionEntity
message UserMentionEntityP{
    int64 id=1;
    string idStr=2;
    repeated int32 indices=3;
    string name=4;
    string screenName=5;
}

//SymbolEntity
message SymbolEntityP{
    repeated int32 indices=1;
    string text=2;
}

//PollEntity
message PollEntityP{
    repeated OptionEntityP options=1;
    string endDatetime=2;
    string durationMinutes=3;
}

//MediaSizesEntity
message MediaSizesEntityP{
    SizeEntityP thumb=1;
    SizeEntityP large=2;
    SizeEntityP medium=3;
    SizeEntityP small=4;
}

//SizeEntity
message SizeEntityP{
    int32 width=1;
    int32 height=2;
    string resize=3;
}

//VariantEntity
message VariantEntityP{
    int64 bitrate=1;
    string contentType=2;
    string url=3;
}
//VideoEntity
message VideoEntityP{
    repeated int32 aspectRatio=1;
    int32 durationMillis=2;
    repeated VariantEntityP variants=3;
}

//AdditionalMediaInfoEntity
message AdditionalMediaInfoEntityP{
    string title=1;
    string description=2;
    bool embeddable=3;
    bool monetizable=4;
}

//OptionEntity
message OptionEntityP{
    int32 position=1;
    string text=2;
}

//MatchingRulesEntity
message MatchingRulesEntityP{
    string tag=1;
    int64 id=2;
    string idStr=3;
}

//Map field
message MapFieldEntry {
    string key = 1;
    bool value = 2;
}

//ExtendedTweetEntity
message ExtendedTweetEntityP{
    string fullText=1;
    repeated int32 displayTextRange=2;
    EntitiesP entities=3;
    ExtendedEntitiesP extendedEntities=4;
}